xc7/utils/prjxray_create_equiv_tiles.py

```diff
@@ -0,0 +1,1253 @@
+
+            pb_types[pb_type_name] = frozenset(pb_types[pb_type_name])
+
+            assert pb_types[pb_type_name] not in site_collections_to_pb_type, (
+                pb_type_name, pb_types[pb_type_name]
+            )
+
+            site_collections_to_pb_type[pb_types[pb_type_name]] = pb_type_name
+
+        pb_types_in_tile = {}
+
+        tiles_that_instance_pb_type = {}
+
+        for tile_type in args.tile_types.split(','):
+            cur.execute(
+                "SELECT pkey FROM tile_type WHERE name = ?", (tile_type, )
+            )
+            tile_type_pkey = cur.fetchone()[0]
+            tile_types[tile_type] = tile_type_pkey
+            node_tile_map[tile_type_pkey] = expand_nodes_in_tile_type(
+                conn, tile_type_pkey
+            )
+            sites[tile_type_pkey], sites_in_tiles[
+                tile_type_pkey] = sites_in_tile_type(conn, tile_type_pkey)
+
+            seen_sets = set()
+            site_type_pkeys = frozenset(sites[tile_type_pkey].keys())
+            seen_sets.add(site_type_pkeys)
+
+            pb_types_in_tile[tile_type_pkey] = [
+                site_collections_to_pb_type[site_type_pkeys]
+            ]
+
+            for site_set in yield_mapped_site_sets(site_remaps,
+                                                   site_type_pkeys):
+                if site_set in seen_sets:
+                    continue
+
+                pb_types_in_tile[tile_type_pkey].append(
+                    site_collections_to_pb_type[site_set]
+                )
+
+            for pb_type in pb_types_in_tile[tile_type_pkey]:
+                if pb_type not in tiles_that_instance_pb_type:
+                    tiles_that_instance_pb_type[pb_type] = []
+
+                tiles_that_instance_pb_type[pb_type].append(tile_type_pkey)
+
+        tile_connections = {}
+
+        for pb_type in tiles_that_instance_pb_type:
+            tile_connections[pb_type] = create_pb_type(
+                conn=conn,
+                pin_assignments=pin_assignments,
+                site_directory=args.site_directory,
+                output_directory=args.output_directory,
+                pb_type=pb_type,
+                site_type_pkeys=pb_types[pb_type],
+                tile_type_pkeys=tiles_that_instance_pb_type[pb_type],
+                site_remaps=site_remaps,
+                site_name_remaps=site_name_remaps,
+                node_tile_map=node_tile_map
+            )
+
+        for tile_type in tile_types:
+            tile_type_pkey = tile_types[tile_type]
+            create_tile(
+                conn=conn,
+                tile_type=tile_type,
+                tile_type_pkey=tile_type_pkey,
+                tile_connections=tile_connections,
+                pb_types=pb_types_in_tile[tile_type_pkey],
+                output_directory=args.output_directory,
+                pin_assignments=pin_assignments,
+            )
+
+
+if __name__ == '__main__':
+    main()
```

Comment:
A lot of this file feels like it might be repeated from other files?

Reply:
Some of it is duplicated, I'll file a cleanup issue for now.

Reply:
https://github.com/SymbiFlow/symbiflow-arch-defs/issues/1089